File: python/moneysocket/beacon/location/websocket.py
# Copyright (c) 2020 Jarret Dyrbye
# Distributed under the MIT software license, see the accompanying
# file LICENSE or http://www.opensource.org/licenses/mit-license.php

from moneysocket.utl.third_party.bolt.tlv import Tlv
from moneysocket.utl.third_party.bolt.namespace import Namespace
from moneysocket.utl.third_party.bolt.bigsize import BigSize


HOST_TLV_TYPE = 0
USE_TLS_TLV_TYPE = 1
USE_TLS_ENUM_VALUE = {0: False,
                      1: True}
PORT_TLV_TYPE = 2
DEFAULT_TLS_PORT = 443
DEFAULT_NO_TLS_PORT = 80

class WebsocketLocation():
    WEBSOCKET_LOCATION_TLV_TYPE = None

    def __init__(self, host, port=None, use_tls=True):
        self.use_tls = use_tls
        self.host = host
        self.port = port if port else (DEFAULT_TLS_PORT if use_tls else
                                       DEFAULT_NO_TLS_PORT)

    def __str__(self):
        return "%s://%s:%s" % ("wss" if self.use_tls else "ws", self.host,
                                self.port)

    def to_dict(self):
        return {'type':    "WebSocket",
                'host':    self.host,
                'port':    self.port,
                'use_tls': self.use_tls}


    @staticmethod
    def from_tlv(tlv):
        assert tlv.t == WebsocketLocation.WEBSOCKET_LOCATION_TLV_TYPE
        tlvs = {tlv.t: tlv for tlv in Namespace.iter_tlvs(tlv.v)}
        if HOST_TLV_TYPE not in tlvs.keys():
            return None, "no host tlv given"
        try:
            host = tlvs[HOST_TLV_TYPE].v.decode("utf8", errors="strict")
        except:
            return None, "error decoding host string"

        if USE_TLS_TLV_TYPE not in tlvs.keys():
            use_tls = True
        else:
            enum_value, remainder, err = BigSize.pop(tlvs[USE_TLS_TLV_TYPE].v)
            if err:
                return None, err
            if enum_value not in USE_TLS_ENUM_VALUE.keys():
                return None, "error decoding use_tls setting"
            use_tls = USE_TLS_ENUM_VALUE[enum_value]

        if PORT_TLV_TYPE not in tlvs.keys():
            port = DEFAULT_TLS_PORT if use_tls else DEFAULT_NO_TLS_PORT
        else:
            port, _, err = BigSize.pop(tlvs[PORT_TLV_TYPE].v)
            if err:
                return None, err
        return WebsocketLocation(host, port=port, use_tls=use_tls), None

    def encode_tlv(self):
        encoded = Tlv(HOST_TLV_TYPE, self.host.encode("utf8")).encode()
        if not self.use_tls:
            encoded += Tlv(USE_TLS_TLV_TYPE, BigSize.encode(0)).encode()
            if self.port != DEFAULT_NO_TLS_PORT:
                encoded += Tlv(PORT_TLV_TYPE,
                               BigSize.encode(self.port)).encode()
        else:
            if self.port != DEFAULT_TLS_PORT:
                encoded += Tlv(PORT_TLV_TYPE,
                               BigSize.encode(self.port)).encode()

        return Tlv(WebsocketLocation.WEBSOCKET_LOCATION_TLV_TYPE,
                   encoded).encode()

File: browser/src/js/moneysocket/beacon/location/websocket.js
// Copyright (c) 2020 Jarret Dyrbye
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php

const BigSize = require('../../utl/bolt/bigsize.js').BigSize;
const Tlv = require('../../utl/bolt/tlv.js').Tlv;
const BinUtl = require('../../utl/bin.js').BinUtl;
const StringUtl = require('../../utl/string.js').StringUtl;



const TLV_TYPE_START = 65979
const WEBSOCKET_LOCATION_TLV_TYPE = TLV_TYPE_START + 8


const HOST_TLV_TYPE = 0;
const USE_TLS_TLV_TYPE = 1;
const USE_TLS_ENUM_VALUE = {0: false,
                            1: true};
const PORT_TLV_TYPE = 2;
const DEFAULT_TLS_PORT = 443;
const DEFAULT_NO_TLS_PORT = 80;

class WebsocketLocation {
    constructor(host, port, use_tls) {
        this.host = host;
        this.use_tls = (use_tls == null) ? true : use_tls;
        this.port = ((port == null) ? (this.use_tls ? DEFAULT_TLS_PORT :
                                                      DEFAULT_NO_TLS_PORT) :
                                       port);
    }

    toDict() {
        return {'type':    "WebSocket",
                'host':    this.host,
                'port':    this.port,
                'use_tls': this.us_tls};
    }

    static fromTlv(tlv) {
        console.assert(tlv.t == WEBSOCKET_LOCATION_TLV_TYPE, "wrong tlv type");
        var tlvs = {};
        Tlv.decodeTlvs(tlv.v).forEach(tlv => {
            tlvs[tlv.t] = tlv;
        });
        if (! (HOST_TLV_TYPE in tlvs)) {
            return [null, "no host tlv given"];
        }
        var host;
        try {
            host = StringUtl.fromUtf8(tlvs[HOST_TLV_TYPE].v);
        }
        catch(err) {
            return [null, "error decoding host string"];
        }
        var use_tls;
        if (! (HOST_TLV_TYPE in tlvs)) {
            use_tls = true;
        } else {
            var [enum_value, remainder, err] = BigSize.pop(
                tlvs[USE_TLS_TLV_TYPE].v);
            if (err != null) {
                return [null, err];
            }
            if (! (enum_value in USE_TLS_ENUM_VALUE)) {
                return [null, "error decoding use_tls setting"];
            }
            use_tls = USE_TLS_ENUM_VALUE[enum_value];
        }
        var port;
        if (! (PORT_TLV_TYPE in tlvs)) {
            port = use_tls ? DEFAULT_TLS_PORT : DEFAULT_NO_TLS_PORT;
        } else {
            var [port, remainder, err] = BigSize.pop(tlvs[PORT_TLV_TYPE].v);
            if (err != null) {
                return [null, err];
            }
        }
        return [new WebsocketLocation(host, port, use_tls), null];
    }

    static encodeTlv() {
        var encoded = Tlv(HOST_TLV_TYPE, StringUtl.toUtf8(host)).encode();

        if (! this.use_tls) {
            encoded += Tlv(USE_TLS_TLV_TYPE, BigSize.encode(0)).encode();
            if (this.port != DEFAULT_NO_TLS_PORT) {
                var encoded_port = Tlv(PORT_TLV_TYPE,
                                       BigSize.encode(this.port)).encode();
                encoded = BinUtl.arrayConcat(encoded, encoded_port);
            }
        } else {
            if (port != DEFAULT_TLS_PORT) {
                var encoded_port = Tlv(PORT_TLV_TYPE,
                                       BigSize.encode(this.port)).encode();
                encoded = BinUtl.arrayConcat(encoded, encoded_port);
            }
        }
        return Tlv(WEBSOCKET_LOCATION_TLV_TYPE, encoded).encode();
    }
}

exports.WebsocketLocation = WebsocketLocation;
